### PR TITLE
sapling: teach gen-deps.py script to update Cargo.lock

### PR DESCRIPTION
**Pull request by strager**
###### Description of changes

Upstream Sapling does not maintain a Cargo.lock file, so Nixpkgs has its own. Teach the gen-deps.py script to update Nixpkgs' Cargo.lock whenever the Cargo.toml files change.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/version-management/sapling/gen-deps.py [hunk 1]
```diff
@@ -1,18 +1,47 @@
...
+    text=True,
+    stdout=subprocess.PIPE,
+).stdout.rstrip().splitlines()
+
+def updateCargoLock():
+    with tempfile.TemporaryDirectory() as tempDir:
+        tempDir = pathlib.Path(tempDir)
+
+        # NOTE(strager): We cannot use shutil.tree because it copies the
+        # read-only permissions.
```
Comment (nagy) [inline pkgs/applications/version-management/sapling/gen-deps.py:33]: Why are we not using `shutil.copytree` followed by a `chmod` on the directory? I have had success with this:
```python
def updateCargoLock():
    with tempfile.TemporaryDirectory() as tempDir:
        tempDir = pathlib.Path(tempDir)
        shutil.copytree(sourceDirectory, tempDir, dirs_exist_ok=True)
        os.chmod(tempDir / "eden" / "scm", 0o755)
        run(["cargo", "fetch"], check=True, cwd=tempDir / "eden" / "scm")
        shutil.copy(tempDir / "eden" / "scm" / "Cargo.lock", "Cargo.lock")
```

File: pkgs/applications/version-management/sapling/gen-deps.py [hunk 2]
```diff
@@ -1,18 +1,47 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i python3 -p "python3.withPackages (ps: with ps; [ requests ])"
+#!nix-shell -i python3 -p cargo -p "python3.withPackages (ps: with ps; [ requests ])"
```
Comment (nagy) [inline pkgs/applications/version-management/sapling/gen-deps.py:2]: I dont think you need that second `-p`

File: pkgs/applications/version-management/sapling/gen-deps.py (hunk 1 above)
Comment (strager) [reply pkgs/applications/version-management/sapling/gen-deps.py:33]: That solution felt nasty to me. Copying with the *correct* permissions seemed better than copying with the *wrong* permissions then fixing them.

**Review by pbar1 — APPROVED** [review]

**Comment by strager** [comment]
Ping. I'd love to get this patch merged.

**Review by thoughtpolice — APPROVED** [review]

**Comment by thoughtpolice** [comment]
lgtm

**Pull request merged**